PocketKit/Sources/PocketKit/Home/HomeViewModel.swift

```diff
@@ -240,30 +194,49 @@ extension HomeViewModel {
             return snapshot
         }
 
-        if let slateLineup = slateLineup,
-           let slates = slateLineup.slates?.compactMap({ $0 as? Slate }) {
-            for slate in slates {
-                guard var recs = slate.recommendations?.compactMap({ $0 as? Recommendation }),
-                      !recs.isEmpty else {
-                    continue
-                }
+        guard let slateSections = self.recomendationsController.sections, !slateSections.isEmpty else {
+            snapshot.appendSections([.loading])
+            snapshot.appendItems([.loading], toSection: .loading)
+            return snapshot
+        }
+
+        for slateSection in slateSections {
+            guard var recommendations = slateSection.objects as? [Recommendation],
+                    !recommendations.isEmpty,
+                  let firstRecommendation = recommendations.first,
+                  let slate = firstRecommendation.slate
+            else {
+                continue
+            }
 
-                let hero = recs.removeFirst()
-                snapshot.appendSections([.slateHero(slate.objectID)])
-                snapshot.appendItems(
-                    [.recommendationHero(hero.objectID)],
-                    toSection: .slateHero(slate.objectID)
-                )
+            let slateId: NSManagedObjectID = slate.objectID
 
-                guard !recs.isEmpty else {
-                    continue
-                }
+            let slateSectionId: Section = .slateHero(slateId)
+
+            let hero = recommendations.removeFirst()
+            if snapshot.indexOfSection(slateSectionId) == nil {
+                snapshot.appendSections([slateSectionId])
+            }
+            snapshot.appendItems(
+                [.recommendationHero(hero.objectID)],
+                toSection: slateSectionId
+            )
+            // Set a rec to always reload so it gets the latest save state. This is because NSFetchedResultsController is not listening for SavedItem changes on recs
+            snapshot.reloadItems([.recommendationHero(hero.objectID)])
+
+            guard !recommendations.isEmpty else {
+                continue
+            }
+
+            if snapshot.indexOfSection(.slateCarousel(slate.objectID)) == nil {
                 snapshot.appendSections([.slateCarousel(slate.objectID)])
-                snapshot.appendItems(
-                    recs[0...min(3, recs.endIndex - 1)].map { .recommendationCarousel($0.objectID) },
-                    toSection: .slateCarousel(slate.objectID)
-                )
             }
+            snapshot.appendItems(
+                recommendations[0...min(3, recommendations.endIndex - 1)].map { .recommendationCarousel($0.objectID) },
```

Comment:
how about `recommendations.prefix(4)` instead? It's safe for sequences shorter than `maxLength` and more readable, wdyt?

Reply:
works for me, i just copied the previous code. i like.